✨ add guide to community health files (#1334)
openjs-foundation/cross-project-council: project-resources/community-health-files-guide.md

# Guide to Community Health Files for OpenJS Foundation Projects

Community health files are documents related to project collaboration. Most commonly, this includes:

- `CODE_OF_CONDUCT.md`
- `CONTRIBUTING.md`
- `GOVERNANCE.md`
- `SECURITY.md`

> [!NOTE]
> See [the full list](https://docs.github.com/en/communities/setting-up-your-project-for-healthy-contributions/creating-a-default-community-health-file#supported-file-types) of supported default community health files

## Default Community Health Files

> Instead of adding these files to every repository within an organization, we can add them only once to a special organization repository which must be named `.github`. GitHub will automatically use the community health files specified at this location by default for all repositories within the organization.

Reviewer: ```suggestion
Instead of adding these files to every repository within a user or organization, we can add them only once to a special user/organization repository which must be named `.github`. GitHub will automatically use the community health files specified at this location by default for all repositories within the user/organization.
```

Author: while this is technically correct -- the best kind of correct -- this guide should probably omit this for clarity.  I don't think any openjs projects live under a user account (and if they do, I suggest they should be moved to an org)

Reviewer: Fair enough, given the audience here.

> [!NOTE]
> All projects must have a LICENSE file, but it is not possible to provide a default LICENSE file for the organization. Every repository requires its own LICENSE file.

#### `CODE_OF_CONDUCT.md`

Organizations/projects should use the following content for their `CODE_OF_CONDUCT.md` file. This will avoid any issues that would arise by duplicating content.

```md
This repository is an OpenJS Foundation project and subscribes to its [Code of Conduct](https://github.com/openjs-foundation/cross-project-council/blob/main/CODE_OF_CONDUCT.md).
```

#### `SECURITY.md`

All organizations/projects must have a security policy. This file must inform users how they can privately report security issues to projects.

At a [bare minimum](https://github.com/ossf/oss-vulnerability-guide/blob/main/templates/security_policies/github_security_policy.md), it should direct users to an email address.

It should also include any other reporting mechanism used by the project, such as [GitHub PVR](https://docs.github.com/en/code-security/security-advisories/working-with-repository-security-advisories/configuring-private-vulnerability-reporting-for-an-organization).

##### Further guidance

> For guidance and assistance on developing security policy, reach out to [the OpenJS Security Collaboration Space](https://github.com/openjs-foundation/security-collab-space/#security-collaboration-space).

Reviewer: perhaps a direct link to https://github.com/openjs-foundation/security-collab-space/blob/main/best-practices-badge.md would also be helpful?

Author: it's definitely helpful. still I hesitate to link it here. we should link to the OpenJS-specific security policy guide once that is in good shape.  the BPB guidance doesn't go into policycraft -- which is good -- but this doc is saying you need a security file/policy, and for those who don't already have a policy, we want to direct them to information that would help them develop it, because what they seek out on their own might lead them to some subpar information.  the BPB guide doesn't add policy information that this guide doesn't mention.  it contains a lot of other great content on BPB but it doesn't seem... wait

I just realized you might have been suggesting this as further guidance in general on best practices, and not specifically further guidance on crafting security policy.  (in this doc, `further guidance` is a sub-topic of the `security.md` topic).  in that case, yeah maybe we add a section for general best practices references like that

what do you think?

Reviewer: Sure, that sounds good.

# Guide to Community Health Files for OpenJS Foundation Projects

> Community health files are documents related to project collaboration. Most commonly, this includes:

Reviewer 2: ```suggestion
Community health files are documents related to project collaboration. Most commonly, these includes:
```

- `CODE_OF_CONDUCT.md`
- `CONTRIBUTING.md`
- `GOVERNANCE.md`

> - `SECURITY.md`

Reviewer 3: We also want to mention CHARTER.md here and say where it should go (depends on whether it's a project with a single repo (in which case it goes in the project) or a project with multiple repos, in which case it must go in a dedicated repo (which ideally should have a common name).

Author: `CHARTER.md` is not a supported file type, so we'll need to add a caveat

Reviewer 3: Right, but _we_ care about projects having such a file and having it in a place that's easy for us to reference.

> > [!NOTE]

Reviewer: Would we ever even want a non-GitHub foundation project?

> > [!NOTE]

Author: the entire document is GitHub-specific.  it's not worth it to start qualifying sections about what is GH only and what it not, unless we have actual usage/requirement for different SCM services.  if/when we have that need, we can update.

> > [!NOTE]

Reviewer 3: > the entire document is GitHub-specific.

So this is interesting, and I think where we need to reach agreement first. My position is that all of those files (README, LICENSE, SECURITY, GOVERNANCE, etc.) are defacto standards, and aren't GitHub-specific. The `.github` repo on the other hand, absolutely is. I'd want us to make a distinction between the two.

> Would we ever even want a non-GitHub foundation project?

Absolutely. Why not?

> > [!NOTE]

Author: I agree with these comments (well, not Jordan's, but that seems like an unnecessary tangent to explore).  My main concern was to avoid making the guide sub-optimal by putting qualifiers everywhere or shifting content.  I expect my changes should address this issue satisfactorily.

> See the [full list](https://docs.github.com/en/communities/setting-up-your-project-for-healthy-contributions/creating-a-default-community-health-file#supported-file-types) of supported default community health files

> ## Default Community Health Files

Reviewer 3: Can we reframe this whole section as: "here's a GitHub-specific thing that is helpful if you have multiple projects:"

Author: - It's not already framed that way?
- It's also still useful for orgs that only have one project

Instead of adding these files to every repository within an organization, we can add them only once to a special organization repository which must be named `.github`. GitHub will automatically use the community health files specified at this location by default for all repositories within the organization.

The filenames must match [what is expected by GitHub](https://docs.github.com/en/communities/setting-up-your-project-for-healthy-contributions/creating-a-default-community-health-file#supported-file-types) and must be in one of these locations:

> - The root of the `.github` repository

Reviewer 3: Let's recommend the best option here. The idea of a best practice guide isn't to inform about all possible options, but to make it easy for folks to do the right thing.

Reviewer 4: I think this should read "in the root of the GitHub repository". I'm not sure what a `.github` repository is.

Reviewer 4: Sorry, nevermind, that's a special repository. Is that a necessity? For instance, while jQuery may have lots of repositories, the core repo is obviously the main one and I'd only want to put these files there.

Reviewer: it's a github repository named `.github`.

Reviewer 4: I think you missed my next comment. I wonder if the wording could change here to account for those that don't need or want a `.github` repository.

Reviewer: It's certainly a necessity for things like a code of conduct to be in every repo - without a `.github` repo, you'd have to repeat them in every repo.

Reviewer 4: 1. That's fine for those with a low number of active repos
2. A code of conduct is required for the organization. The absence of the file in one of the repos does not mean the organization doesn't prescribe to a code of conduct, especially if the main repo has it.

Reviewer: I suppose that's technically true, but participants in one of the other repos won't be able to become aware of the CoC if it's not in one of the github-prescribed locations for it. iow, there's no "main" repo in a generic sense in a github org, so a user wouldn't necessarily have a way of knowing which one it is.

Reviewer: either way, these are guidelines, not mandates, so it's fine if jquery doesn't want to follow them - in general, orgs really should have a .github repo.

Reviewer 4: I still see some edge cases where this reads weird, but I'll set this up for jQuery

Reviewer 4: Regarding the original ask, I think the best option is top-level.

Author: > those that don't need or want a .github repository.

a major goal of this guide is that all orgs should have/use a `.github` repo

Reviewer 3: > a major goal of this guide is that all orgs should have/use a .github repo

I agree that this is the easiest way to get at the outcome we care about, which is that every project has all of the right files in the right places.

I think framing it that way helps understand the benefits of the `.github` repo (which I don't think are clear to everyone): it acts kind of like a dynamic repo template system (or like properties of the prototype of an JS object, to use an analogy that should speak more clearly to folks in this crowd).

- The `.github` folder in the `.github` repository
- The `docs` folder in the `.github` repository

Where possible, it is best to have only one file for each type of community health file, specified in the `.github` repository, and to avoid having individual files for each repository. However, specific project needs may require repository-specific files.

> [!Important]
> If a file also exists in a repository, then that file will take precedence over the organization default.

### Community Health Files That Projects Should Have

- Organizations/projects must have at least `CODE_OF_CONDUCT.md` and `SECURITY.md` files.
- A [`CONTRIBUTING.md`](https://docs.github.com/en/communities/setting-up-your-project-for-healthy-contributions/setting-guidelines-for-repository-contributors) file is highly recommended for all organizations/projects.

> - A `GOVERNANCE.md` file should be provided for all organizations/projects with formal governance.

Reviewer 5: ```suggestion
- A `GOVERNANCE.md` file should be provided for all organizations/projects with formal governance. It is a requirement for all OpenJS projects.
```

Reviewer 3: Currently, it's only required for Impact projects (see https://github.com/openjs-foundation/cross-project-council/blob/main/PROJECT_PROGRESSION.md#onboarding-checklist).

Author: This is a good point.  Although we should avoid duplicating requirements whose Source of Truth™ is elsewhere.  If governance is required for impact projects, then by extension a `GOVERNANCE.md` file is required, but this document doesn't need to spell that out.

Author: I deleted the comment, but you may have gotten an email notification where I commented and pushed back against this.  On further consideration, I agree it is helpful to specify where files are required for certain projects.

Reviewer 4: FWIW, there are some Impact projects that don't have a GOVERNANCE.md.

Author: Tobie is right -- this is not the place for policy changes.  Let's capture what the reality is right now, which is that it's required for impact projects.

> [!NOTE]

> > All projects must have a LICENSE file, but it is not possible to provide a default LICENSE file for the organization. Every repository requires its own LICENSE file.

Reviewer 5: ```suggestion
> All projects must have a LICENSE file, but it is not possible to provide a default LICENSE file for the organization. Every repository requires its own LICENSE file. This file __must__ be included in any released artifact. 
```

Author: This is a good point, but this document isn't about `LICENSE` files at all; this note is included so people don't make the mistake of thinking they can specify an org-wide license to apply to all repos.  Requirements for IP, licensing, bundling, etc. are outside the scope of this document.

#### `CODE_OF_CONDUCT.md`

Organizations/projects should use the following content for their `CODE_OF_CONDUCT.md` file. This will avoid any issues that would arise by duplicating content.

```md

> This repository is an OpenJS Foundation project and subscribes to its [Code of Conduct](https://github.com/openjs-foundation/cross-project-council/blob/main/CODE_OF_CONDUCT.md).

Reviewer 3: ```suggestion
This projects follows the OpenJS Foundation Code of Conduct available at [code-of-conduct.openjsf.org](https://code-of-conduct.openjsf.org/).
```

Reviewer 4: I think you meant to change another word there. "This projects"

Reviewer 3: Thank you!

> This repository is an OpenJS Foundation project and subscribes to its [Code of Conduct](https://github.com/openjs-foundation/cross-project-council/blob/main/CODE_OF_CONDUCT.md).

Reviewer 4: ```suggestion
All projects in this organization are OpenJS Foundation projects and subscribe to its [Code of Conduct](https://github.com/openjs-foundation/cross-project-council/blob/main/CODE_OF_CONDUCT.md).
```

?

Author: I adopted Tobie's alternate suggested change for this.
